src/index.tsx

```diff
@@ -138,23 +140,36 @@ const VisGraph = forwardRef<
     if (isEqual(graph.nodes, prevNodes.current)) {
       return; // No change!
     }
-    const { added, removed, updated } = diff(prevNodes.current, graph.nodes);
+    const { added, removed, updated } = diff(nodes.get(), graph.nodes);
 
-    nodes.remove(removed);
-    nodes.add(added);
-    nodes.update(updated);
+    if (removed.length) {
+      nodes.remove(removed);
+    }
+    // Shallow clone nodes to ensure props aren't mutated!
+    if (added.length) {
+      nodes.add(shallowClone(added));
+    }
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
```

Comment:
Same as analogical section in the update edges useEffect

Reply:
> I will run it tomorrow, I have some concerns about a few things but I don't know if we wanna adress them in this mr.
> 
> * IMO tracking the previous values of nodes and edges should be at least done in seperate useEffect, the best solution would be to abstract pretty simple `usePreviousValue` hook.
> * Two useEffects relating tracking the updated, added, deleted nodes basically do the same thing and should also be abstracted into some other custom hook
> 
> Other than that, in terms of code it is fine for me

I do see your point about setting up a custom hook to do the functionality that the edges/nodes need, I'll probably tie in the refs and other bits into that custom hook.

As for tracking the previous values of the nodes and edges, given that we are only using them here, we'd essentially be using an effect hook with the same dependencies to do that tracking, and given how small that functionality is, it's a bit wasteful in this specialized use case, especially since I need the ref to exist before this effect, but change the value of it afterwards, which unfortunately makes a `usePreviousValue` hook a bit more complicated to set up.